File: examples/ccsds/aem_to_trajectory.rs
//! Parse an AEM and convert a segment to an AttitudeTrajectory for interpolation.

#[allow(unused_imports)]
use brahe as bh;
use brahe::ccsds::AEM;
use brahe::traits::{AttitudeProvider, Trajectory};

fn main() {
    bh::initialize_eop().unwrap();

    // Parse an AEM with two quaternion segments
    let aem = AEM::from_file("test_assets/ccsds/aem/AEMExampleG4.txt").unwrap();

    // Segment 1 carries no INTERPOLATION_METHOD, so it converts cleanly to
    // the default slerp trajectory. Segment 0 sets INTERPOLATION_METHOD =
    // HERMITE, which has no AttitudeTrajectory equivalent and would error.
    let traj = aem.segment_to_attitude_trajectory(1).unwrap();
    println!("Trajectory: {} states", traj.len());
    println!("  Frame A:       {}", traj.frame_a);
    println!("  Frame B:       {}", traj.frame_b);
    println!("  Interpolation: {:?}", traj.interpolation_method);
    println!("  Has rates:     {}", traj.has_rates());
    println!("  Start:         {}", traj.start_epoch().unwrap());
    println!("  End:           {}", traj.end_epoch().unwrap());

    // Slerp-query the attitude at the midpoint of the trajectory's span
    let t0 = traj.start_epoch().unwrap();
    let t1 = traj.end_epoch().unwrap();
    let mid = t0 + (t1 - t0) / 2.0;
    let quaternion = traj.quaternion(mid).unwrap();
    let wire = quaternion.to_vector(false);
    println!(
        "\nInterpolated quaternion [Q1, Q2, Q3, QC] at {}: [{:.5}, {:.5}, {:.5}, {:.5}]",
        mid, wire[0], wire[1], wire[2], wire[3]
    );
}

File: examples/ccsds/aem_to_trajectory.py
# /// script
# dependencies = ["brahe"]
# ///
"""
Parse an AEM and convert a segment to an AttitudeTrajectory for interpolation.
"""

import brahe as bh
from brahe.ccsds import AEM

bh.initialize_eop()

# Parse an AEM with two quaternion segments
aem = AEM.from_file("test_assets/ccsds/aem/AEMExampleG4.txt")

# Segment 1 carries no INTERPOLATION_METHOD, so it converts cleanly to the
# default slerp trajectory. Segment 0 sets INTERPOLATION_METHOD = HERMITE,
# which has no AttitudeTrajectory equivalent and would raise an error.
traj = aem.segment_to_attitude_trajectory(1)
print(f"Trajectory: {len(traj)} states")
print(f"  Frame A:       {traj.frame_a}")
print(f"  Frame B:       {traj.frame_b}")
print(f"  Interpolation: {traj.interpolation_method}")
print(f"  Has rates:     {traj.has_rates}")
print(f"  Start:         {traj.start_epoch}")
print(f"  End:           {traj.end_epoch}")

# Slerp-query the attitude at the midpoint of the trajectory's span
t0 = traj.start_epoch
t1 = traj.end_epoch
mid = t0 + (t1 - t0) / 2.0
quaternion = traj.quaternion(mid)
wire = quaternion.to_vector(scalar_first=False)
print(
    f"\nInterpolated quaternion [Q1, Q2, Q3, QC] at {mid}: "
    f"[{wire[0]:.5f}, {wire[1]:.5f}, {wire[2]:.5f}, {wire[3]:.5f}]"
)
